[BE-Feature] 개인 일정 동기화 구현

kwon204 opened the pull request:
<!-- 제목 템플릿
[파트명-Feat] [파트명-Refactor] [파트명-Fix] [파트명-Chore] [파트명-Remove]
-->
## #️⃣ 연관된 이슈>
- #357 
- #354 
## 📝 작업 내용> 이번 PR에서 작업한 내용을 간략히 설명해주세요(이미지 첨부 가능)
- long polling api를 구현했습니다. 
  - 타임아웃은 1분으로 설정했습니다. 
  - <s>타임아웃 전에 이벤트가 발생하면 "sync"를 응답합니다. </s>
  - 타임아웃 전에 이벤트가 발생하면 변경된 일정 데이터를 응답합니다.
  - 타임아웃이 일어나면 "timeout"을 응답합니다.
  - <s>"sync" 응답을 받으면 syncToken과 함께 get events 요청을 하여 변경된 이벤트를 받습니다.</s>
- <s>동기화 시 필요한 syncToken을 구현했습니다. </s>
  - <s>토큰에 저장된 시간데이터를 사용하여 마지막 동기화 시각 이후에 변경된 데이터를 응답합니다. </s>
 

## 🙏 여기는 꼭 봐주세요! > 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요
- 삭제된 이벤트에 대한 동작이 부실하다고 생각되는데 좋은 의견이 있다면 공유하면 좋을 거 같습니다!
- => 삭제된 일정에 대해서는 다음의 응답이 전송됩니다.
```json
{
	"googleEventId": "id",
	"titile": "title",
	"startDateTime":  "2025-04-06:12:00:00",
	"endDateTime": "2025-04-06:12:10:00" ,
	"status": "cancelled"
}
```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a new endpoint to enable real-time synchronization of personal events for more responsive updates.
  - Added a new record for synchronization responses, enhancing the structure of event data handling.

- **Refactor**
  - Streamlined integration of external calendar events with improved asynchronous processing.
  - Replaced previous real-time subscription functionality with a more efficient notification mechanism, simplifying event updates.

- **Bug Fixes**
  - Adjusted method implementations for consistency and clarity in handling personal events.

- **Tests**
  - Updated test cases to align with the refined event synchronization logic.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kwon204 committed:
feat: long polling 구현

kwon204 committed:
feat: syncToken 구현

마지막 동기화 시작 포함

kwon204 committed:
feat: syncToken을 사용한 동기화와 long polling 응답 구현

kwon204 committed:
chore: 주석 삭제 및 swagger 추가

efdao commented on backend/src/main/java/endolphin/backend/domain/personal_event/PersonalEventController.java:
```diff
@@ -50,11 +57,12 @@ public class PersonalEventController {
+    public ResponseEntity<SyncPersonalEvent> getPersonalEvents(
+        @Valid @RequestParam LocalDate startDate,
+        @Valid @RequestParam LocalDate endDate,
+        @RequestParam String syncToken) {
```
p1; 최초 요청 시 syncToken을 전달받지 않는다면 required = false 설정해야 할 것 같습니다.

efdao commented on backend/src/main/java/endolphin/backend/global/util/SyncTokenUtil.java:
```diff
@@ -0,0 +1,21 @@
+        return Base64.getEncoder().encodeToString(tokenData.getBytes(StandardCharsets.UTF_8));
+    }
+
+    public static LocalDateTime getLastSyncTime(String token) {
+        byte[] decodedBytes = Base64.getDecoder().decode(token);
+        String decoded = new String(decodedBytes, StandardCharsets.UTF_8);
+
+        return LocalDateTime.parse(decoded);
+    }
```
p1; 토큰 검증, 에러 처리가 필요해보입니다.

efdao commented on backend/src/main/java/endolphin/backend/domain/personal_event/PersonalEventService.java:
```diff
@@ -247,87 +266,87 @@ private void validatePersonalEventUser(PersonalEvent personalEvent, User user) {
+                    changedDates.add(personalEvent.getStartTime().toLocalDate());
+                    changedDates.add(personalEvent.getEndTime().toLocalDate());
+                    deletePersonalEvent(personalEvent, user, discussions);
+                });
```
just ask; 코드스타일 구글 xml로 적용되어있는거 맞나요?

kwon204 replied:
아 데스크탑에는 적용되어있지 않았네요. 적용하겠습니다.

efdao reviewed (changes requested):
현재 구조가 클라이언트가 sync 응답을 받고 get events를 한 후 반복문으로 변경사항을 확인한 후 렌더링하게되는건가요?

kwon204 committed:
feat: DeferredResultManager 추가

- Long polling에 사용되는 비동기 응답 객체 관리 기능
- DeferredResult를 생성하고 저장
- 이벤트 발생 시 응답 값을 DeferredResult에 설정하고 응답
- timeout 설정

kwon204 commented:
> 현재 구조가 클라이언트가 sync 응답을 받고 get events를 한 후 반복문으로 변경사항을 확인한 후 렌더링하게되는건가요?

네 맞습니다. 그런데 다시 생각해보니 syncToken 없이도 할 수 있을 거 같아서 새로 수정 후 올리겠습니다

kwon204 committed:
feat: long polling을 위한 DeferredResult 객체 관리 클래스 추가

kwon204 committed:
refactor: syncToken 삭제 및 리팩토링

kwon204 committed:
test: 테스트 코드 수정

kwon204 committed:
fix: timeout 시간 6초 -> 60초로 변경

kwon204 committed:
refactor: 사용하지 않는 클래스 삭제

kwon204 committed:
fix: deferredResult가 setResult() 되지 않는 버그 수정

efdao commented on backend/src/main/java/endolphin/backend/domain/personal_event/PersonalEventService.java:
```diff
@@ -220,24 +219,19 @@ public void preprocessPersonalEvents(User user, Discussion discussion) {
+                case CONFIRMED -> upsertPersonalEventByGoogleEvent(
+                    googleEvent, discussions, user, googleCalendarId);
+                case CANCELLED -> deletePersonalEventByGoogleEvent(
+                    googleEvent, discussions, user, googleCalendarId);
```
just ask; 제 3의 상태는 들어올 일 없나요?

kwon204 replied:
tentive는 예전에 논의한 결과, 무시하기로 했습니다

efdao commented on backend/src/main/java/endolphin/backend/domain/personal_event/dto/SyncPersonalEvent.java:
```diff
@@ -0,0 +1,17 @@
+    LocalDateTime startDateTime,
+    LocalDateTime endDateTime,
+    String status
+) {
```
p3; 메소드명을 통일하는 것이 좋을 것 같습니다. 저는 of()로 작성했습니다.

kwon204 replied:
https://docs.oracle.com/javase/tutorial/datetime/overview/naming.html 
이렇게 convention이 존재했었네요.. 이렇게 맞추는 건 어떤가요?

efdao replied:
좋습니다

efdao commented on backend/src/main/java/endolphin/backend/global/util/DeferredResultManager.java:
```diff
@@ -0,0 +1,47 @@
+        new ConcurrentHashMap<>();
+    private final static Long TIMEOUT = 60000L;
+
+    public DeferredResult<ResponseEntity<SyncResponse>> create(User user) {
```
p3; 연결이 존재하는 User에 대해 create()가 호출되면 어떻게되나요?

kwon204 replied:
기존의 연결은 끊고 새로운 연결을 만들도록 수정하였습니다.

efdao reviewed (changes requested):
코멘트 확인 부탁드립니다.

kwon204 committed:
fix: 반복 요청을 하면 이전의 연결은 끊고 새로운 연결만 유지하도록 수정

kwon204 committed:
refactor: long polling 연결 여부 확인 로직 변경

efdao reviewed (approved):
고생하셨습니다
